var/spack/repos/builtin/packages/gcc/package.py

```diff
@@ -264,6 +267,108 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
+        for vargs in ('-dumpfullversion', '-dumpversion'):
+            try:
+                output = spack.compiler.get_compiler_version_output(exe, vargs)
+                match = version_regex.search(output)
+                if match:
+                    return match.group(1)
+            except spack.util.executable.ProcessError:
+                pass
+            except Exception as e:
+                tty.debug(e)
+
+        return None
+
+    @classmethod
+    def determine_variants(cls, exes, version_str):
+        spec_str = 'gcc@{0} languages={1}'
+        languages, compilers = set(), {}
+        for exe in exes:
+            if 'gcc' in exe:
+                languages.add('c')
+                compilers['c'] = exe
+            if 'g++' in exe:
+                languages.add('c++')
+                compilers['cxx'] = exe
+            if 'gfortran' in exe:
+                languages.add('fortran')
+                compilers['fortran'] = exe
+
+        spec_str = spec_str.format(version_str, ','.join(languages))
+        return [Spec.from_detection(
+            spec_str, extra_attributes={'compilers': compilers}
+        )]
+
+    @classmethod
+    def validate_detected_spec(cls, spec, extra_attributes):
+        # For GCC 'compilers' is a mandatory attribute
+        msg = ('the extra attribute "compilers" must be set for '
+               'the detected spec "{0}"'.format(spec))
+        assert 'compilers' in extra_attributes, msg
+
+        compilers = extra_attributes['compilers']
+        for constraint, key in {
+            'languages=c': 'c',
+            'languages=c++': 'cxx',
+            'languages=fortran': 'fortran'
+        }.items():
+            if spec.satisfies(constraint, strict=True):
+                msg = '{0} not in {1}'
+                assert key in compilers, msg.format(key, spec)
+
+    @property
+    def cc(self):
+        msg = "cannot retrieve C compiler [spec is not concrete]"
+        assert self.spec.concrete, msg
+        if self.spec.external:
+            return self.spec.extra_attributes['compilers'].get('c', None)
+        return self.spec.prefix.bin.gcc if 'languages=c' in self.spec else None
+
+    @property
+    def cxx(self):
+        msg = "cannot retrieve C++ compiler [spec is not concrete]"
+        assert self.spec.concrete, msg
+        if self.spec.external:
+            return self.spec.extra_attributes['compilers'].get('cxx', None)
+        result = None
+        if 'languages=c++' in self.spec:
+            result = os.path.join(self.spec.prefix.bin, 'g++')
+        return result
+
+    @property
+    def fortran(self):
+        msg = "cannot retrieve Fortran compiler [spec is not concrete]"
+        assert self.spec.concrete, msg
+        if self.spec.external:
+            return self.spec.extra_attributes['compilers'].get('fortran', None)
+        result = None
+        if 'languages=fortran' in self.spec:
+            result = self.spec.prefix.bin.gfortran
+        return result
```

Comment:
This is pretty simple, but repetitive. It works properly:
```console
Spack version 0.14.2
Python 3.7.4, Linux x86_64
>>> import spack.spec
>>> s = spack.spec.Spec('gcc')
>>> s.concretize()
>>> s.package.cc
'/usr/bin/x86_64-linux-gnu-gcc-9'
```
but maybe we can further simplify the logic in a following PR.